File: 0001_0100/0047/0047.py
with open("input.txt", "r") as f, open("output.txt", "w") as q:
    n = int(f.read())
    d = n
    s = sum(int(x) for x in str(n))
    for i in (x for x in range(2, n // 2 + 1) if n % x == 0):
        t = sum(int(x) for x in str(i))
        if t > s or t == s and i < d:
            d, s = i, t
    q.write(str(d))

File: 0001_0100/0047/0047.go
package main

import (
	"fmt"
	"os"
)

func sum(n int) int {
	s := 0
	for n > 0 {
		s += n % 10
		n /= 10
	}
	return s
}

func main() {
	f, _ := os.Open("input.txt")
	q, _ := os.Create("output.txt")
	var n int
	fmt.Fscan(f, &n)
	d, s := 1, 1
	for i := 2; i <= n/2; i++ {
		if n%i == 0 && sum(i) > s {
			d, s = i, sum(i)
		}
	}
	if sum(n) > s {
		d = n
	}
	fmt.Fprint(q, d)
}
